sdks/python/apache_beam/runners/portability/job_server.py

```diff
@@ -20,18 +20,143 @@
+import tempfile
+import threading
 import time
-from subprocess import Popen
-from subprocess import check_output
-from threading import Lock
 
+import grpc
 
-class DockerizedJobServer(object):
+from apache_beam.portability.api import beam_job_api_pb2_grpc
+from apache_beam.runners.portability import local_job_service
+
+
+class JobServer(object):
+  def start(self):
+    """Starts this JobServer, returning a grpc service to which to submit jobs.
+    """
+    raise NotImplementedError()
+
+  def stop(self):
+    """Stops this job server."""
+    raise NotImplementedError()
+
+
+class ExternalJobServer(JobServer):
+  def __init__(self, endpoint):
+    self._endpoint = endpoint
+
+  def start(self):
+    channel = grpc.insecure_channel(self._endpoint)
+    grpc.channel_ready_future(channel).result()
+    return beam_job_api_pb2_grpc.JobServiceStub(channel)
+
+  def stop(self):
+    pass
+
+
+class EmbeddedJobServer(JobServer):
+  def start(self):
+    return local_job_service.LocalJobServicer()
+
+  def stop(self):
+    pass
+
+
+class PersistentJobServer(JobServer):
+  """Wraps a JobServer such that its stop will automatically be called on exit.
   """
-   Spins up the JobServer in a docker container for local execution
+  def __init__(self, job_server):
+    self._lock = threading.Lock()
+    self._job_server = job_server
+    self._started = False
+
+  def start(self):
+    with self._lock:
+      if not self._started:
+        self._endpoint = self._job_server.start()
+        self._started = True
+        atexit.register(self.stop)
+        signal.signal(signal.SIGINT, self.stop)
+    return self._endpoint
+
+  def stop(self):
+    with self._lock:
+      if self._started:
+        self._job_server.stop()
+      self._started = False
+
+
+class SubprocessJobServer(JobServer):
+  """An abstract base class for JobServers run as an external process."""
+  def __init__(self):
+    self._process_lock = threading.RLock()
+    self._process = None
+    self._local_temp_root = None
+
+  def subprocess_cmd_and_endpoint(self):
+    raise NotImplementedError()
+
+  def start(self):
+    with self._process_lock:
+      if self._process:
+        self.stop()
+      cmd, endpoint = self.subprocess_cmd_and_endpoint()
+      logging.debug("Starting job service with %s", cmd)
+      try:
+        self._process = subprocess.Popen([str(arg) for arg in cmd])
+        self._local_temp_root = tempfile.mkdtemp(prefix='beam-temp')
+        wait_secs = .1
+        channel = grpc.insecure_channel(endpoint)
+        channel_ready = grpc.channel_ready_future(channel)
+        while True:
+          if self._process.poll() is not None:
+            logging.error("Starting job service with %s", cmd)
+            raise RuntimeError(
+                'Job service failed to start up with error %s' %
+                self._process.poll())
+          try:
+            channel_ready.result(timeout=wait_secs)
+            break
+          except grpc.FutureTimeoutError:
```

Comment:
Might have to add `grpc.RpcError` here to catch `grpc._channel._Rendezvous` which frequently happens on docker on mac.